File: src/algorithms/trees/bst-operations/bst-range-sum-iterative/bst-range-sum-iterative.test.ts
import { describe, it, expect } from "vitest";
import { bstRangeSumIterative } from "./sources/bst-range-sum-iterative.ts?fn";

interface BSTNode {
  value: number;
  left: BSTNode | null;
  right: BSTNode | null;
}
function node(value: number, left: BSTNode | null = null, right: BSTNode | null = null): BSTNode {
  return { value, left, right };
}
const tree = node(4, node(2, node(1), node(3)), node(6, node(5), node(7)));

describe("bstRangeSumIterative", () => {
  it("sums values in [3, 7]", () => {
    expect(bstRangeSumIterative(tree, 3, 7)).toBe(3 + 4 + 5 + 6 + 7);
  });
  it("sums all values", () => {
    expect(bstRangeSumIterative(tree, 1, 7)).toBe(28);
  });
  it("returns 0 for no match", () => {
    expect(bstRangeSumIterative(tree, 10, 20)).toBe(0);
  });
  it("handles null tree", () => {
    expect(bstRangeSumIterative(null, 1, 7)).toBe(0);
  });
});

File: src/algorithms/trees/bst-operations/bst-range-sum-iterative/__tests__/BSTRangeSumIterative_test.cpp
// g++ -o bst_rsi_test BSTRangeSumIterative_test.cpp && ./bst_rsi_test
#include "../sources/BSTRangeSumIterative.cpp"
#include <cassert>
#include <iostream>

BSTNode* makeRSIterNode(int value, BSTNode* left = nullptr, BSTNode* right = nullptr) {
    BSTNode* node = new BSTNode(value);
    node->left = left;
    node->right = right;
    return node;
}

int main() {
    BSTNode* tree = makeRSIterNode(4, makeRSIterNode(2, makeRSIterNode(1), makeRSIterNode(3)), makeRSIterNode(6, makeRSIterNode(5), makeRSIterNode(7)));

    assert(bstRangeSumIterative(tree, 3, 7) == 25);
    assert(bstRangeSumIterative(tree, 1, 7) == 28);
    assert(bstRangeSumIterative(tree, 10, 20) == 0);
    assert(bstRangeSumIterative(nullptr, 1, 7) == 0);

    std::cout << "All tests passed!" << std::endl;
    return 0;
}
